src/command/archive/common.c

```diff
@@ -249,6 +249,12 @@ archiveAsyncExec(ArchiveMode archiveMode, const StringList *commandExec)
         // Detach from parent process
         forkDetach();
 
+        // Close any open file descriptors above the standard three (stdin, stdout, stderr). Don't check the return value since we
+        // don't know which file descriptors are actually open (might be none). It's possible that there are open files >= 1024 but
+        // there is no easy way to detect that and this should give us enough descriptors to do our work.
+        for (int fd = 3; fd < 1024; fd++)
+            close(fd);
```

Comment:
I'm a little confused. If we inherit file descriptors from Postgresql then are we in danger here of closing FDs actually in use? Also I don't see any tests for this. And:

"Not checking the return value when closing the file may lead to silent loss of data. This can especially be observed with NFS and with disk quota.

A successful close does not guarantee that the data has been successfully saved to disk, as the kernel defers writes. It is not common for a file system to flush the buffers when the stream is closed. If you need to be sure that the data is physically stored use fsync(2). (It will depend on the disk hardware at this point.) "

Reply:
> If we inherit file descriptors from Postgresql then are we in danger here of closing FDs actually in use?

Yeah, we might close FDs in use but *we* are not using them, Postgres is. The FD will still be open in Postgres.

> Not checking the return value when closing the file may lead to silent loss of data.

Postgres is responsible for syncing and closing these files.

> Also I don't see any tests for this

We are testing it in `command/archive-get` and `command/archive-push` at least to the extent that it is not breaking anything. We don't have a good test to make sure that the code does not disappear -- it's not clear to me how we would do that since it is happening two forks down. Even if the code disappeared it wouldn't exactly represent a bug on our part. We are just trying to compensate for PostgreSQL not closing file handles before exec'ing pgbackrest.

Reply:
Note that, technically, there's two independent sets of file descriptors after the fork() happens, which is why it doesn't make any different to PG that we close *our* copy of the file descriptors that PG had open when we were fork()'d from PG.